### DISCUSSION TITLE
Learning the graph structure for link prediction

### DISCUSSION THREAD

**Discussion by GabbySuwichaya** [pyg-team/pytorch_geometric#2681] [Q&A] [answered]
Hi, I am working on the application of classification/link prediction. So, far I have looked at the code such as SuperGAT.  I have noticed is that the dataset used in SuperGAT (e.g., Citations) seems to already have the graph structure (the set of edges to be considered in training and testing). 

However, in my application, all I have is the data attributes .....without the edges or adjacent matrix or mask to identify which nodes may be connected to which. 

My questions are 

1. Do you have any recommended way to estimate the edges/adjacent matrix?  I am looking for time-efficient methods to solve the problem as well?   
2. I am very new to GAT.... But do GAT normally requires the edges of the graph in computing the attention? 
3. Will GAT work better if it considers all the possible pairs of edges or a restricted set of edges? 
4. I have tried to extend SuperGAT to connect all the possible pairs, but the training results are not so good. So, how other works normally handle the training of all the possible pairs? 
5. Also, just to make sure.... is my observation correct ... did I misunderstand anything?

**Comment by rusty1s** [comment, marked as answer]
Thanks for your interest.

1. In general, a GNN operates on a pre-defined graph-structure. We assume that a graph is given, and learn from features and structures in an end-to-end fashion in order to derive at some predictions. In case you do not have any graph-structure, one can also operate on the fully-connected graph (in that case `GAT` models a Transformer model). As such, attention scores can be interpreted as learning a soft adjacency matrix per layer. However, keep in mind that this package is more suited towards sparse graph-structured data rather than fully-connected graphs.
2. Yes, `GAT` computes an attention score per edge.
3. This depends. As mentioned earlier, considering all possible pairs of edges will yield a Transformer model. GAT can be seen as a sparse and local variant of such models.
4. I don't think extending SuperGAT for fully-connected graphs makes a lot of sense, since the link prediction objective can not sample negative node pairs.

**Reply by GabbySuwichaya** [reply]
I see. Thank you very much. This gives me lots more understanding. 

I would like to ask a few more questions. 

Could you help explain why the transformer can work well even without the edge connections? I am quite new to both transformer and GAT... I thnk of GAT as a process in Transformer... 
However, when I try to train with fully connected edges like the transformer, I find that the learning progress is quite slow.... So, I would like to understand the underlying mechanism that makes a transformer works well even without the edge structure .....

**Reply by rusty1s** [reply]
The fully-connected transformer just attends to all other nodes, i.e., it tries to find nodes that are important for the central node itself to make more accurate predictions. In that sense, the transformer tries to learn a graph structure by itself. I think this is the most reasonable approach to tackle this problem in case you do not have further knowledge about node relationships. However, since the Transformer has to learn everything from scratch (without any inductive bias), it will take longer to train and needs a lot more data.

**Reply by GabbySuwichaya** [reply]
@rusty1s  Thanks so much.... Your explanation does help me a lot.  I would like to ask one last question ....   
Do people use any positional encoder with GAT? Or they just use it with the transformer due to the lack of knowledge about node relationships?

**Reply by rusty1s** [reply]
In general, one does not use positional encoding in GAT/GNNs since there is no notion of "position" in (most) graphs. There exists some work on this topic though, see, e.g., [here](https://arxiv.org/abs/2010.02863).

**Reply by shahinghasemi** [reply]
> The fully-connected transformer just attends to all other nodes, i.e., it tries to find nodes that are important for the central node itself to make more accurate predictions. In that sense, the transformer tries to learn a graph structure by itself. I think this is the most reasonable approach to tackle this problem in case you do not have further knowledge about node relationships. However, since the Transformer has to learn everything from scratch (without any inductive bias), it will take longer to train and needs a lot more data.

I have the same problem. Does calculating the adjacency matrix using some similarity algorithms e.g. `Jaccard` or `Cosine` make sense? Let's say we have a binary feature matrix `X` of (n, f) then we calculate `adj = Jaccard(X)` now `adj` is (n, n) which somehow simulates the relations between pairs of nodes.

**Reply by zcaicaros** [reply]
> > The fully-connected transformer just attends to all other nodes, i.e., it tries to find nodes that are important for the central node itself to make more accurate predictions. In that sense, the transformer tries to learn a graph structure by itself. I think this is the most reasonable approach to tackle this problem in case you do not have further knowledge about node relationships. However, since the Transformer has to learn everything from scratch (without any inductive bias), it will take longer to train and needs a lot more data.
> 
> I have the same problem. Does calculating the adjacency matrix using some similarity algorithms e.g. `Jaccard` or `Cosine` make sense? Let's say we have a binary feature matrix `X` of (n, f) then we calculate `adj = Jaccard(X)` now `adj` is (n, n) which somehow simulates the relations between pairs of nodes.

As mentioned by @rusty1s , I guess the best way is to use the transformer to learn a soft adjacent matrix, such that we can utilize the power of neural nets to explore the whole adj space to find the most suitable one. I guess this will be stronger than a hard-coded one, such as Jaccard or Cosine?

**Comment by taghizad3h** [comment]
If you have some prior knowledge about the structure of adjacency matrix e.g sparsity, smoothness of input, etc. you can use these unsupervised objectives to learn the initial graph. For more details refer to this paper [Exploring Structure-Adaptive Graph Learning for Robust Semi-Supervised Classification](https://arxiv.org/abs/1904.10146).
